include/nds/rsa.h

```diff
@@ -14,23 +20,77 @@ extern "C" {
 
 #include <nds/ndstypes.h>
 
+/// Context that holds information about the RSA heap
 typedef struct swiRSAHeapContext
 {
-    void *heapStart;
-    void *heapEnd;
-    size_t heapSize;
+    void *heapStart __attribute__((aligned(4)));    ///< Start of the heap rounded up to 4-byte boundary
+    void *heapEnd __attribute__((aligned(4)));      ///< start + size rounded down to 4-byte boundary
+    u32 heapSize;    ///< Heap size matched to the above rounded values
 } swiRSAHeapContext_t;
 
+/// Struture that holds pointers to the RSA buffers
 typedef struct swiRSAbuffers
 {
-    void *dst;
-    const void *sig;
-    const void *key;
+    void *dst;  ///< Pointer to the output buffer
+    const void *sig;    ///< Pointer to the signature buffer
+    const void *key;    ///< Pointer to the RSA key buffer
 } swiRSAbuffers_t;
 
-int swiRSAInitHeap(swiRSAHeapContext_t *ctx, void *heapStart, size_t heapSize);
-int swiRSADecryptRAW(swiRSAHeapContext_t *ctx, swiRSAbuffers_t *rsabuffers, size_t len);
+/// Initialize RSA Heap
+///
+/// @param ctx 12-byte heap information structure that gets set with heap
+///            start, heap end and heap length.
+/// @param heapStart Pointer to the start of the heap.
+/// @param heapSize Size of the heap in bytes (should be usually 4096).
+void swiRSAInitHeap(swiRSAHeapContext_t *ctx, void *heapStart, u32 heapSize);
+
+/// Decrypt data without unpadding (aside from stripping leading 00h bytes;
+/// most (or all) signatures are containing one leading 00h byte, so the
+/// returned @ref lenDst value will be usually 7Fh).
+///
+/// @param ctx RSA Heap context
+/// @param rsabuffers Struct containing pointers to destination buffer,
+///                   signature and public key to use.
+/// @param lenDst Pointer to memory where to store the length of the
+///               destination buffer.
+/// @return 1 if success, 0 if failed.
+int swiRSADecryptRAW(swiRSAHeapContext_t *ctx, swiRSAbuffers_t *rsabuffers, u32 *lenDst);
+
+/// Decrypt buffer and additionally remove the "01h,min eight FFh,00h"
+/// padding.  The output at @ref dst can be theoretically max 80h-bytes (or
+/// shorter due to removed padding). In practice, the DSi is often using only
+/// the first 14h-bytes at dst (aka the last 14h-bytes from sig) as SHA1 or
+/// SHA1-HMAC value (RSA-SHA1).
+///
+/// @param ctx RSA Heap context
+/// @param dst Pointer to output buffer (should be 128 bytes).
+/// @param sig Pointer to signature buffer (should be 128 bytes).
+/// @param key Pointer to key buffer (should be 128 bytes).
+/// @return 1 if success, 0 if failed
 int swiRSADecrypt(swiRSAHeapContext_t *ctx, void *dst, const void *sig, const void *key);
+
+/// Decrypt signature data and do some extra processing for extracting a
```

Comment:
This is copy-pasted material from GBATEK. As surprising as it may seem, GBATEK is not openly licensed (especially given at least its 3DS section is itself largely copy-pastes), and thus copying long snippets from it into BlocksDS is not acceptable.

Reply:
Sorry, forgot to check about license issues. Will fix